### ISSUE TITLE
Prevent "soft" memory leaks by suggesting to shrink

### ISSUE THREAD

**Issue by Rudxain** [rust-lang/rust-clippy#10079]
### What it does

`Vec`s, `String`s, and other heap-allocated types, don't automatically shrink their `capacity` when the `len()` is small enough. This lint acts as a reminder to shrink whenever there's a call to `remove` 

### Lint Name

remove_without_shrink

### Category

perf, restriction

### Advantage

- reminds to free unused memory
- good for beginners and experienced devs alike, who are unaware of the lack of auto-shrink, or have too much cognitive load (respectively)

### Drawbacks

- too many false positives, unless it's smarter (i.e. only triggering if multiple elements are removed at once)

### Example

```rust
vector.remove(0);
```

Could be written as:

```rust
vector.remove(0);
vector.shrink_to_fit();
```

**Comment by matthiaskrgr** [comment]
Would love to see some benchmarks because I can hardly imagine this being worthwhile when removing a single item.

**Comment by Alexendoo** [comment]
It makes the most sense to shrink a collection when it has just had a large amount of stuff removed from, will no longer have things added and will be kept around for a long time after the removal. That's a very difficult set of circumstances to detect

**Comment by Rudxain** [comment]
> Would love to see some benchmarks

That's the reason why I included the "restriction" category, it doesn't performance enough, only memory use

> `remove` inside loops and suggest shrinking outside

I totally agree! that's much better

> That's a very difficult set of circumstances to detect

That's unfortunate. So we have 2 options:
1. Too many false positives, easy implementation, "restriction" category
2. Least possible false positives, some false negatives, very complicated implementation, "perf" category

I vote for # 1. But # 2 could also be paired with a new lint, `shrink_in_loop`, which triggers whenever  any shrink* is called in loop

**Comment by Rudxain** [comment]
[Recent discussion about this](https://internals.rust-lang.org/t/vec-is-asymmetric-with-memory-handling/22449)
